staging/src/k8s.io/apimachinery/pkg/util/net/http.go

```diff
@@ -206,52 +206,67 @@ func GetHTTPClient(req *http.Request) string {
 	return "unknown"
 }
 
-// SourceIPs splits the comma separated X-Forwarded-For header or returns the X-Real-Ip header or req.RemoteAddr,
-// in that order, ignoring invalid IPs. It returns nil if all of these are empty or invalid.
+// SourceIPs splits the comma separated X-Forwarded-For header and joins it with
+// the X-Real-Ip header and/or req.RemoteAddr, ignoring invalid IPs.
+// The X-Real-Ip is omitted if it's already present in the X-Forwarded-For chain.
+// The req.RemoteAddr is always the last IP in the returned list.
+// It returns nil if all of these are empty or invalid.
 func SourceIPs(req *http.Request) []net.IP {
+	var srcIPs []net.IP
+
 	hdr := req.Header
 	// First check the X-Forwarded-For header for requests via proxy.
 	hdrForwardedFor := hdr.Get("X-Forwarded-For")
-	forwardedForIPs := []net.IP{}
 	if hdrForwardedFor != "" {
 		// X-Forwarded-For can be a csv of IPs in case of multiple proxies.
 		// Use the first valid one.
 		parts := strings.Split(hdrForwardedFor, ",")
 		for _, part := range parts {
 			ip := net.ParseIP(strings.TrimSpace(part))
 			if ip != nil {
-				forwardedForIPs = append(forwardedForIPs, ip)
+				srcIPs = append(srcIPs, ip)
 			}
 		}
 	}
-	if len(forwardedForIPs) > 0 {
-		return forwardedForIPs
-	}
 
 	// Try the X-Real-Ip header.
 	hdrRealIp := hdr.Get("X-Real-Ip")
 	if hdrRealIp != "" {
 		ip := net.ParseIP(hdrRealIp)
-		if ip != nil {
-			return []net.IP{ip}
+		// Only append the X-Real-Ip if it's not already contained in the X-Forwarded-For chain.
+		if ip != nil && !containsIP(srcIPs, ip) {
+			srcIPs = append(srcIPs, ip)
```

Comment:
Should x-real-ip  be prepended?

Reply:
Probably, except that would change the logic of ClientIP. Maybe the solution is to just break the dependency of ClientIP on SourceIPs, what do you think?

Reply:
Doesn't prepending it preserve ClientIP behavior? Before this change, it is always returned by ClientIP if present. To maintain that behavior, it needs to be prepended.

Reply:
See:

https://github.com/kubernetes/kubernetes/blob/a9c7529f990159ef327716e7c61984d5b7197115/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L260-L266

Reply:
| X-Forwarded-For | X-Real-IP | GetClientIP original | GetClientIP appended | GetClientIP prepended|
|--|--|--|--|--|
| true | true | first x-forwarded-for ip | first x-forwarded-for ip | x-real-ip |
| true | false | first x-forwarded-for ip | first x-forwarded-for ip | first x-forwarded-for ip |
| false | true | x-real-ip | x-real-ip | x-real-ip |
| false | false | remote ip | remote ip | remote ip |

Please double check my logic, but I think in the case when both the x-forwarded-for & x-real-ip is provided, the x-real-ip needs to be appended to be consistent with the original GetClientIP behavior.

Reply:
Ah ya, you are right. Thanks for the truth table :). I think I was overlooking the early return of forwarded for.